### PR TITLE
Class parameters support

### PR DESCRIPTION
**Pull request by kozdra**
This pull request aims to provide support for parametrized classes.  This is based on the ongoing effort on classes and parameters in general, which has progressed very much lately.  As much code is reused as possible without introducing confusing virtual multiple inheritance to the codebase, treating class references (and classes themselves) similar to cells (and modules, respectively).

The changes are split into several (hopefully) well-thought commits for easier review and possibility of integrating the changes gradually, each explaining its meaning in detail.

### CODE REVIEW

File: src/V3LinkDot.cpp [hunk 1]
```diff
@@ -1986,6 +1986,24 @@ class LinkDotResolveVisitor final : public VNVisitor {
             refp->user5p(nodep);
         }
     }
+    bool paramSkip(const AstNode* nodep) {
```
Comment (kbieganski) [inline src/V3LinkDot.cpp:1989]: This function could use some comments as to why it's necessary. Also maybe a slightly more descriptive name?

File: src/V3LinkDot.cpp [hunk 2]
```diff
@@ -2516,13 +2536,25 @@ class LinkDotResolveVisitor final : public VNVisitor {
...
+                    "class reference parameter not removed by V3Param");
+        VL_RESTORER(m_ds);
+        VL_RESTORER(m_pinSymp);
+        {
+            // ClassRef's have pins, so track
+            if (nodep->classOrPackagep())
+                m_pinSymp = m_statep->getNodeSym(nodep->classOrPackagep());
+            m_ds.init(m_curSymp);
+            UINFO(4, "(Backto) Link ClassOrPackageRef: " << nodep << endl);
+            // if (debug()) nodep->dumpTree(cout, "linkcell:");
```
Comment (kbieganski) [inline src/V3LinkDot.cpp:2554]: I'm aware Verilator has a fair amount of commented out code, but I think we shouldn't add any more :)

File: src/V3Param.cpp [hunk 3]
```diff
@@ -825,6 +825,129 @@ class ParamProcessor final {
         // v3Global.rootp()->dumpTreeFile(v3Global.debugFilename("param-out.tree"));
     }
 
+    AstNodeModule* classRefDeparam(AstClassRefDType* nodep, AstNodeModule* modp,
```
Comment (kbieganski) [inline src/V3Param.cpp:828]: All three `*Deparam` methods have a fair bit of duplicated code.

File: src/V3Param.cpp [hunk 4]
```diff
@@ -965,11 +1102,15 @@ class ParamVisitor final : public VNVisitor {
     }
 
     virtual void visit(AstClassRefDType* nodep) override {
-        if (nodep->paramsp()) {
-            nodep->paramsp()->v3warn(E_UNSUPPORTED, "Unsupported: parameterized classes");
-            pushDeletep(nodep->paramsp()->unlinkFrBackWithNext());
-        }
-        iterateChildren(nodep);
+        string* const genHierNamep = new string(m_generateHierName);
```
Comment (kbieganski) [inline src/V3Param.cpp:1105]: Not sure about this. Could we put this string in a struct together with the class ref in `m_classRefps`? I don't think there's any advantage to using `user*` here.

**Review by kbieganski — CHANGES_REQUESTED** [review]
I think it looks pretty good. Please write up a short description of your changes.

File: src/V3Param.cpp (hunk 4 above)
Comment (kozdra) [reply src/V3Param.cpp:1105]: This is basically doing the same thing that `visit(AstCell*)` did, and I am not sure why it happens exactly.
The two lists could probably get merged together.

File: src/V3LinkDot.cpp [hunk 5]
```diff
@@ -1987,6 +1987,18 @@ class LinkDotResolveVisitor final : public VNVisitor {
         }
     }
 
+    bool isParamedClassRef(const AstNode* nodep) {
+        if (const auto* classRefp = VN_CAST(nodep, ClassOrPackageRef)) {
+            if (classRefp->paramsp()) return true;
+
+            const auto* classp = classRefp->classOrPackageNodep();
+            while (const auto* typedefp = VN_CAST(classp, Typedef)) classp = typedefp->subDTypep();
+            if (VN_IS(classp, ClassRefDType) && VN_AS(classp, ClassRefDType)->paramsp())
```
Comment (kbieganski) [inline src/V3LinkDot.cpp:1996]:
```suggestion
            return VN_IS(classp, ClassRefDType) && VN_AS(classp, ClassRefDType)->paramsp();
```

File: src/V3Param.cpp [hunk 6]
```diff
@@ -758,8 +758,24 @@ class ParamProcessor final {
         }
     }
 
+    void cellApplyDeparam(AstNode* nodep, AstNodeModule* paramedModp) {
```
Comment (kbieganski) [inline src/V3Param.cpp:761]: `cellDeparam` and `cellApplyDeparam` no longer only deal with cells. Let's get rid of the `cell` in their names. Also, `deparam` vs `applyDeparam` doesn't tell me much about the roles of these functions relative to each other. Can we find a more descriptive verb?

File: src/V3Param.cpp [hunk 7]
```diff
@@ -769,57 +785,61 @@ class ParamProcessor final {
...
         // Must be a separate loop, as constant conversion may have changed some pointers.
         // if (debug()) nodep->dumpTree(cout, "-cel2: ");
         string longname = srcModp->name() + "_";
         bool any_overrides = false;
-        // Must always clone __Vrcm (recursive modules)
-        if (nodep->recursive()) any_overrides = true;
-        if (debug() > 8 && nodep->paramsp())
-            nodep->paramsp()->dumpTreeAndNext(cout, "-cellparams: ");
+        AstPin* paramsp = nullptr;
+        if (const auto* cellp = VN_CAST(nodep, Cell)) {
```
Comment (kbieganski) [inline src/V3Param.cpp:796]: Wouldn't it be more readable to have separate a `deparam` function for each case and then the common bits in other functions we call into? That would reduce the ifology a bit.

File: src/V3Width.cpp [hunk 8]
```diff
@@ -2437,6 +2437,17 @@ class WidthVisitor final : public VNVisitor {
...
+    virtual void visit(AstClassOrPackageRef* nodep) override {
+        if (nodep->didWidthAndSet()) return;
+        userIterateChildren(nodep, nullptr);
+    }
+    virtual void visit(AstDot* nodep) override {
+        // We can only reach this from constify called during V3Param (so before linkDotParam)
+        // ... #(Cls#(...)::...) ...
+        //                ^^~~~ this is our DOT
+        nodep->v3warn(E_UNSUPPORTED,
+                      "dotted expressions in parameters; you can work this around with a typedef");
```
Comment (kbieganski) [inline src/V3Width.cpp:2449]:
```suggestion
                      "dotted expressions in parameters" << warnMore << "Suggest use a typedef");
```

**Review by kbieganski — CHANGES_REQUESTED** [review]

File: src/V3Param.cpp (hunk 7 above)
Comment (kozdra) [reply src/V3Param.cpp:796]: It is a bit complicated.  We need a bunch of ifs anyway in order to call the correct function in the first place, and then every function would need to do take 6+ arguments, as the generic logic is quite tightly interleaved with the specialized logic.  I will push an example idea to class-param-rework when I am finished with addressing these comments.

**Review by kbieganski — APPROVED** [review]

**Pull request closed without merging**